pkg/cacheutil/memcached_client.go

```diff
@@ -0,0 +1,448 @@
+	// to the max concurrency we have.
+	c.workers.Add(c.config.MaxGetMultiBatchConcurrency)
+	for i := 0; i < c.config.MaxGetMultiBatchConcurrency; i++ {
+		go c.getMultiQueueProcessLoop()
+	}
+
+	return c, nil
+}
+
+func (c *memcachedClient) Stop() {
+	close(c.stop)
+
+	// Wait until all workers have terminated.
+	c.workers.Wait()
+}
+
+func (c *memcachedClient) SetAsync(ctx context.Context, key string, value []byte, ttl time.Duration) error {
+	return c.enqueueAsync(func() {
+		start := time.Now()
+		c.operations.WithLabelValues(opSet).Inc()
+
+		span, _ := tracing.StartSpan(ctx, "memcached_set")
+		err := c.client.Set(&memcache.Item{
+			Key:        key,
+			Value:      value,
+			Expiration: int32(time.Now().Add(ttl).Unix()),
+		})
+		span.Finish()
+		if err != nil {
+			c.failures.WithLabelValues(opSet).Inc()
+			level.Warn(c.logger).Log("msg", fmt.Sprintf("failed to store item with key %s to memcached", key), "err", err)
+			return
+		}
+
+		c.duration.WithLabelValues(opSet).Observe(time.Since(start).Seconds())
+	})
+}
+
+func (c *memcachedClient) GetMulti(ctx context.Context, keys []string) map[string][]byte {
+	batches, err := c.getMultiBatched(ctx, keys)
+	if err != nil {
+		level.Warn(c.logger).Log("msg", "failed to fetch keys from memcached", "err", err)
+
+		// In case we have both results and an error, it means some batch requests
+		// failed and other succeeded. In this case we prefer to log it and move on,
+		// given returning some results from the cache is better than returning
+		// nothing.
+		if len(batches) == 0 {
+			return nil
+		}
+	}
+
+	hits := map[string][]byte{}
+	for _, items := range batches {
+		for key, item := range items {
+			hits[key] = item.Value
+		}
+	}
+
+	return hits
+}
+
+func (c *memcachedClient) getMultiBatched(ctx context.Context, keys []string) ([]map[string]*memcache.Item, error) {
+	// Do not batch if the input keys are less then the max batch size.
+	if (c.config.MaxGetMultiBatchSize <= 0) || (len(keys) <= c.config.MaxGetMultiBatchSize) {
+		items, err := c.getMultiSingle(ctx, keys)
+		if err != nil {
+			return nil, err
+		}
+
+		return []map[string]*memcache.Item{items}, nil
+	}
+
+	// Calculate the number of expected results.
+	batchSize := c.config.MaxGetMultiBatchSize
+	numResults := len(keys) / batchSize
+	if len(keys)%batchSize != 0 {
+		numResults++
+	}
+
+	// Split input keys into batches and schedule a job for it.
+	results := make(chan *memcachedGetMultiResult, numResults)
+	defer close(results)
+
+	go func() {
+		for batchStart := 0; batchStart < len(keys); batchStart += batchSize {
+			batchEnd := batchStart + batchSize
+			if batchEnd > len(keys) {
+				batchEnd = len(keys)
+			}
+
+			c.getMultiQueue <- &memcachedGetMultiBatch{
+				ctx:     ctx,
+				keys:    keys[batchStart:batchEnd],
+				results: results,
+			}
+		}
+	}()
+
+	// Wait for all batch results. In case of error, we keep
+	// track of the last error occurred.
+	items := make([]map[string]*memcache.Item, 0, numResults)
```

Comment:
Maybe `sync.Pool` could be used here?

Reply:
Maybe, but fine to play with it with some microbenchmark on later PRs.